src/WiFiManager.cpp

```diff
@@ -1,228 +1,206 @@
 #include "WiFiManager.h"
 
 #include "CaptivePortal.h"
+#include "Mappers/EspWiFiTypesMapper.h"
 #include "VisualStateManager.h"
+#include "WiFiCredentials.h"
+#include "WiFiScanManager.h"
 
 #include <ArduinoJson.h>
 #include <LittleFS.h>
 #include <WiFi.h>
 
+#include <esp_wifi_types.h>
+
 #include <vector>
 
+#include <nonstd/span.hpp>
+
 const char* const TAG = "WiFiManager";
 
 using namespace OpenShock;
 
-struct WifiCredentials {
-  String ssid;
-  String password;
-  std::uint16_t wifiIndex;
-  std::uint8_t attempts;
-};
-static std::vector<WifiCredentials> s_wifiCredentials;
-static WiFiState s_wifiState;
-
-void SetWiFiState(WiFiState state) {
-  s_wifiState = state;
-  VisualStateManager::SetWiFiState(state);
+void _broadcastWifiAddNetworkSuccess(const char* ssid) {
+  DynamicJsonDocument doc(64);
+  doc["type"]    = "wifi";
+  doc["subject"] = "add_network";
+  doc["status"]  = "success";
+  doc["ssid"]    = ssid;
+  CaptivePortal::BroadcastMessageJSON(doc);
+}
+void _broadcastWifiAddNetworkError(const char* error) {
+  DynamicJsonDocument doc(64);
+  doc["type"]    = "wifi";
+  doc["subject"] = "add_network";
+  doc["status"]  = "error";
+  doc["error"]   = error;
+  CaptivePortal::BroadcastMessageJSON(doc);
 }
 
-bool SaveCredentials() {
-  File file = LittleFS.open("/networks", FILE_WRITE);
-  if (!file) {
-    ESP_LOGE(TAG, "Failed to open networks file for writing");
-    return false;
-  }
+struct WiFiNetwork {
+  char ssid[33];
+  std::uint8_t bssid[6];
+  std::uint8_t channel;
+  std::int8_t rssi;
+  wifi_auth_mode_t authmode;
+  std::uint16_t reconnectionCount;
+  std::uint8_t credentialsId;
+};
 
-  DynamicJsonDocument doc(1024);
-  JsonArray networks = doc.createNestedArray("networks");
+static std::vector<WiFiNetwork> s_wifiNetworks;
+static std::vector<WiFiCredentials> s_wifiCredentials;
 
+bool _addNetwork(const char* ssid, const String& password) {
+  // Bitmask representing available credential IDs (0-31)
+  std::uint32_t bits = 0;
   for (auto& cred : s_wifiCredentials) {
-    JsonObject network  = networks.createNestedObject();
-    network["ssid"]     = cred.ssid;
-    network["password"] = cred.password;
-  }
+    if (strcmp(cred.ssid().data(), ssid) == 0) {
+      ESP_LOGE(TAG, "Failed to add WiFi credentials: credentials for %s already exist", ssid);
+      cred.setPassword(password);
+      cred.save();
+      return true;
+    }
 
-  if (serializeJson(doc, file) == 0) {
-    ESP_LOGE(TAG, "Failed to serialize networks file");
-    file.close();
-    return false;
+    // Mark the credential ID as used
+    bits |= 1u << cred.id();
```

Comment:
Would be great if you could macro these bit shift operations:
- `SET_USED_FLAG(x)`
- `CLEAR_USED_FLAG(x)` (if that happens at all, ever)
- `IS_USED_FLAG_SET(x)`
I think it would serve readability well.

Reply:
Wont do this, might add docs later